File: Services/SubtitleParser.vb
Imports System.Text
Imports System.Text.RegularExpressions
Imports Subtitle_draft_GMTPC.Models

Namespace Services
    ''' <summary>
    ''' Service phân tích cú pháp phụ đề SRT và ASS
    ''' </summary>
    Public Class SubtitleParser

        ''' <summary>
        ''' Phát hiện định dạng phụ đề từ nội dung
        ''' </summary>
        Public Shared Function DetectFormat(content As String) As SubtitleFormat
            If String.IsNullOrWhiteSpace(content) Then Return SubtitleFormat.Unknown

            Dim trimmed = content.Trim()

            ' ASS thường có [Script Info] hoặc [Events] hoặc dòng Dialogue:
            If trimmed.Contains("[Script Info]") OrElse trimmed.Contains("[Events]") OrElse trimmed.Contains("Dialogue:") Then
                Return SubtitleFormat.Ass
            End If

            ' SRT thường có pattern: so --> so
            If Regex.IsMatch(trimmed, "\d{1,2}:\d{2}:\d{2}[,.]\d{2,3}\s*-->\s*\d{1,2}:\d{2}:\d{2}[,.]\d{2,3}") Then
                Return SubtitleFormat.Srt
            End If

            Return SubtitleFormat.Unknown
        End Function

        ''' <summary>
        ''' Parse nội dung phụ đề thành danh sách các dòng
        ''' </summary>
        Public Shared Function Parse(content As String) As List(Of SubtitleLine)
            Dim format = DetectFormat(content)
            Select Case format
                Case SubtitleFormat.Ass
                    Return ParseAss(content)
                Case SubtitleFormat.Srt
                    Return ParseSrt(content)
                Case Else
                    Return New List(Of SubtitleLine)()
            End Select
        End Function

        ''' <summary>
        ''' Parse phụ đề ASS
        ''' </summary>
        Public Shared Function ParseAss(content As String) As List(Of SubtitleLine)
            Dim result = New List(Of SubtitleLine)()
            If String.IsNullOrWhiteSpace(content) Then Return result

            Dim lines = content.Split({Environment.NewLine, vbCr, vbLf}, StringSplitOptions.None)

            For Each line In lines
                Dim trimmed = line.Trim()
                ' Chỉ lấy các dòng Dialogue
                If trimmed.StartsWith("Dialogue:") Then
                    Try
                        Dim assLine = New AssSubtitleLine(line)
                        result.Add(assLine)
                    Catch
                        ' Bo qua dong loi
                    End Try
                End If
            Next

            Return result
        End Function

        ''' <summary>
        ''' Parse phụ đề SRT
        ''' </summary>
        Public Shared Function ParseSrt(content As String) As List(Of SubtitleLine)
            Dim result = New List(Of SubtitleLine)()
            If String.IsNullOrWhiteSpace(content) Then Return result

            ' Chuan hoa line endings
            Dim normalized = content.Replace(vbCr & vbLf, vbLf).Replace(vbCr, vbLf)
            Dim lines = normalized.Split(vbLf)

            Dim i As Integer = 0
            While i < lines.Length
                Dim line = lines(i).Trim()

                ' Bo qua dong trong
                If String.IsNullOrWhiteSpace(line) Then
                    i += 1
                    Continue While
                End If

                ' Thu parse index (so thu tu)
                Dim index As Integer = 0
                If Integer.TryParse(line, index) Then
                    ' Dong tiep theo phai la time line
                    If i + 1 < lines.Length Then
                        Dim timeLine = lines(i + 1).Trim()
                        Dim timeMatch = Regex.Match(timeLine, "(.+?)\s*-->\s*(.+)")

                        If timeMatch.Success Then
                            Dim startTime = SubtitleLine.ParseSrtTime(timeMatch.Groups(1).Value)
                            Dim endTime = SubtitleLine.ParseSrtTime(timeMatch.Groups(2).Value)

                            ' Gom cac dong text tiep theo cho den khi gap dong trong
                            Dim textLines = New List(Of String)()
                            i += 2
                            While i < lines.Length AndAlso Not String.IsNullOrWhiteSpace(lines(i).Trim())
                                textLines.Add(lines(i).TrimEnd())
                                i += 1
                            End While

                            Dim text = String.Join(Environment.NewLine, textLines)
                            Dim srtLine = New SrtSubtitleLine(index, startTime, endTime, text)
                            srtLine.OriginalText = String.Join(Environment.NewLine, {index.ToString(), timeLine, text})
                            srtLine.Content = srtLine.OriginalText
                            result.Add(srtLine)
                        Else
                            i += 1
                        End If
                    Else
                        i += 1
                    End If
                Else
                    i += 1
                End If
            End While

            Return result
        End Function

        ''' <summary>
        ''' Xuat danh sach SubtitleLine thanh noi dung text hoan chinh
        ''' </summary>
        Public Shared Function ToText(lines As List(Of SubtitleLine), format As SubtitleFormat) As String
            If lines Is Nothing OrElse lines.Count = 0 Then Return String.Empty

            If format = SubtitleFormat.Srt Then
                Return ToSrtText(lines)
            Else
                Return ToAssText(lines)
            End If
        End Function

        ''' <summary>
        ''' Xuat ra dinh dang SRT
        ''' </summary>
        Private Shared Function ToSrtText(lines As List(Of SubtitleLine)) As String
            Dim sb = New StringBuilder()
            Dim index = 1

            For Each line In lines
                Dim srtLine = TryCast(line, SrtSubtitleLine)
                If srtLine IsNot Nothing Then
                    srtLine.Index = index
                    sb.AppendLine(srtLine.RebuildLine())
                    sb.AppendLine()
                    index += 1
                Else
                    ' Neu la AssSubtitleLine, chuyen doi sang format SRT
                    Dim timeLine = String.Format("{0} --> {1}", SubtitleLine.FormatSrtTime(line.StartTime), SubtitleLine.FormatSrtTime(line.EndTime))
                    sb.AppendLine(index.ToString())
                    sb.AppendLine(timeLine)

                    Dim assLine = TryCast(line, AssSubtitleLine)
                    If assLine IsNot Nothing Then
                        sb.AppendLine(assLine.DialogText)
                    Else
                        sb.AppendLine(line.OriginalText)
                    End If
                    sb.AppendLine()
                    index += 1
                End If
            Next

            Return sb.ToString().TrimEnd()
        End Function

        ''' <summary>
        ''' Xuat ra dinh dang ASS - chi xuat cac dong Dialogue, khong xuat header
        ''' </summary>
        Private Shared Function ToAssText(lines As List(Of SubtitleLine)) As String
            Dim sb = New StringBuilder()

            For Each line In lines
                Dim assLine = TryCast(line, AssSubtitleLine)
                If assLine IsNot Nothing Then
                    sb.AppendLine(assLine.RebuildLine())
                Else
                    ' Neu la SrtSubtitleLine, chuyen doi sang format ASS
                    Dim srtLine = TryCast(line, SrtSubtitleLine)
                    If srtLine IsNot Nothing Then
                        Dim dialogueLine = String.Format("Dialogue: 0,{0},{1},Default,,0000,0000,0000,,{2}",
                            SubtitleLine.FormatAssTime(line.StartTime),
                            SubtitleLine.FormatAssTime(line.EndTime),
                            srtLine.Text.Replace(Environment.NewLine, "\N"))
                        sb.AppendLine(dialogueLine)
                    End If
                End If
            Next

            Return sb.ToString().TrimEnd()
        End Function

        ''' <summary>
        ''' Xuat danh sach SubtitleLine thanh noi dung text giu nguyen format goc
        ''' </summary>
        Public Shared Function ToOriginalText(lines As List(Of SubtitleLine), format As SubtitleFormat) As String
            Return ToText(lines, format)
        End Function
    End Class
End Namespace

File: Services/SubtitleParser.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;
using Subtitle_draft_GMTPC.Models;

namespace Subtitle_draft_GMTPC.Services
{
    /// <summary>
    /// Service phân tích cú pháp phụ đề SRT và ASS
    /// </summary>
    public class SubtitleParser
    {
        /// <summary>
        /// Làm sạch text: loại bỏ BOM, zero-width chars, ký tự điều khiển
        /// Cần thiết khi copy từ trình duyệt (Chrome/Edge) hoặc các nguồn khác
        /// </summary>
        public static string SanitizeContent(string content)
        {
            if (string.IsNullOrEmpty(content)) return content;

            var sb = new StringBuilder(content.Length);
            foreach (var ch in content)
            {
                var code = (int)ch;
                // Giữ lại các ký tự bình thường
                if (code == 0x9) { // Tab
                    sb.Append(ch);
                }
                else if (code == 0xA) { // Line feed
                    sb.Append(ch);
                }
                else if (code == 0xD) { // Carriage return
                    sb.Append(ch);
                }
                else if (code >= 0x20 && code <= 0xFFFF) { // Ký tự in được
                    // Loại bỏ BOM và zero-width chars
                    if (code != 0xFEFF && code != 0x200B && code != 0x200C && code != 0x200D && code != 0x200E && code != 0x200F && code != 0x202A && code != 0x202B && code != 0x202C && code != 0x202D && code != 0x202E && code != 0xFEFF && code != 0xFFFC)
                    {
                        sb.Append(ch);
                    }
                }
                else if (code > 0xFFFF) { // Emoji, ký tự đặc biệt > U+FFFF
                    sb.Append(ch);
                }
                // Các ký tự điều khiển khác (< 0x20) bị bỏ qua
            }
            return sb.ToString();
        }

        /// <summary>
        /// Phát hiện định dạng phụ đề từ nội dung
        /// </summary>
        public static SubtitleFormat DetectFormat(string content)
        {
            if (string.IsNullOrWhiteSpace(content)) return SubtitleFormat.Unknown;

            var trimmed = content.Trim();

            // ASS thường có [Script Info] hoặc [Events] hoặc dòng Dialogue:
            if (trimmed.Contains("[Script Info]") || trimmed.Contains("[Events]") || trimmed.Contains("Dialogue:"))
            {
                return SubtitleFormat.Ass;
            }

            // SRT thường có pattern: số --> số
            if (Regex.IsMatch(trimmed, @"\d{1,2}:\d{2}:\d{2}[,.]\d{2,3}\s*-->\s*\d{1,2}:\d{2}:\d{2}[,.]\d{2,3}"))
            {
                return SubtitleFormat.Srt;
            }

            return SubtitleFormat.Unknown;
        }

        /// <summary>
        /// Parse nội dung phụ đề thành danh sách các dòng
        /// </summary>
        public static List<SubtitleLine> Parse(string content)
        {
            var format = DetectFormat(content);
            switch (format)
            {
                case SubtitleFormat.Ass:
                    return ParseAss(content);
                case SubtitleFormat.Srt:
                    return ParseSrt(content);
                default:
                    return new List<SubtitleLine>();
            }
        }

        /// <summary>
        /// Parse phụ đề ASS
        /// </summary>
        public static List<SubtitleLine> ParseAss(string content)
        {
            var result = new List<SubtitleLine>();
            if (string.IsNullOrWhiteSpace(content)) return result;

            var lines = content.Split(new[] { Environment.NewLine, "\r", "\n" }, StringSplitOptions.None);

            foreach (var line in lines)
            {
                var trimmed = line.Trim();
                // Chỉ lấy các dòng Dialogue
                if (trimmed.StartsWith("Dialogue:"))
                {
                    try
                    {
                        var assLine = new AssSubtitleLine(line);
                        result.Add(assLine);
                    }
                    catch
                    {
                        // Bo qua dong loi
                    }
                }
            }

            return result;
        }

        /// <summary>
        /// Parse phụ đề SRT
        /// </summary>
        public static List<SubtitleLine> ParseSrt(string content)
        {
            var result = new List<SubtitleLine>();
            if (string.IsNullOrWhiteSpace(content)) return result;

            // Chuan hoa line endings
            var normalized = content.Replace("\r\n", "\n").Replace("\r", "\n");
            var lines = normalized.Split('\n');

            int i = 0;
            while (i < lines.Length)
            {
                var line = lines[i].Trim();

                // Bo qua dong trong
                if (string.IsNullOrWhiteSpace(line))
                {
                    i++;
                    continue;
                }

                // Thu parse index (so thu tu)
                if (int.TryParse(line, out int index))
                {
                    // Dong tiep theo phai la time line
                    if (i + 1 < lines.Length)
                    {
                        var timeLine = lines[i + 1].Trim();
                        var timeMatch = Regex.Match(timeLine, @"(.+?)\s*-->\s*(.+)");

                        if (timeMatch.Success)
                        {
                            var startTime = SubtitleLine.ParseSrtTime(timeMatch.Groups[1].Value);
                            var endTime = SubtitleLine.ParseSrtTime(timeMatch.Groups[2].Value);

                            // Gom cac dong text tiep theo cho den khi gap dong trong
                            var textLines = new List<string>();
                            i += 2;
                            while (i < lines.Length && !string.IsNullOrWhiteSpace(lines[i].Trim()))
                            {
                                textLines.Add(lines[i].TrimEnd());
                                i++;
                            }

                            var text = string.Join(Environment.NewLine, textLines);
                            var srtLine = new SrtSubtitleLine(index, startTime, endTime, text);
                            srtLine.OriginalText = string.Join(Environment.NewLine, new[] { index.ToString(), timeLine, text });
                            srtLine.Content = srtLine.OriginalText;
                            result.Add(srtLine);
                        }
                        else
                        {
                            i++;
                        }
                    }
                    else
                    {
                        i++;
                    }
                }
                else
                {
                    i++;
                }
            }

            return result;
        }

        /// <summary>
        /// Xuat danh sach SubtitleLine thanh noi dung text hoan chinh
        /// </summary>
        public static string ToText(List<SubtitleLine> lines, SubtitleFormat format)
        {
            if (lines == null || lines.Count == 0) return string.Empty;

            if (format == SubtitleFormat.Srt)
            {
                return ToSrtText(lines);
            }
            else
            {
                return ToAssText(lines);
            }
        }

        /// <summary>
        /// Xuat ra dinh dang SRT
        /// </summary>
        private static string ToSrtText(List<SubtitleLine> lines)
        {
            var sb = new StringBuilder();
            var index = 1;

            foreach (var line in lines)
            {
                var srtLine = line as SrtSubtitleLine;
                if (srtLine != null)
                {
                    srtLine.Index = index;
                    sb.AppendLine(srtLine.RebuildLine());
                    sb.AppendLine();
                    index++;
                }
                else
                {
                    // Neu la AssSubtitleLine, chuyen doi sang format SRT
                    var timeLine = $"{SubtitleLine.FormatSrtTime(line.StartTime)} --> {SubtitleLine.FormatSrtTime(line.EndTime)}";
                    sb.AppendLine(index.ToString());
                    sb.AppendLine(timeLine);

                    var assLine = line as AssSubtitleLine;
                    if (assLine != null)
                    {
                        sb.AppendLine(assLine.DialogText);
                    }
                    else
                    {
                        sb.AppendLine(line.OriginalText);
                    }
                    sb.AppendLine();
                    index++;
                }
            }

            return sb.ToString().TrimEnd();
        }

        /// <summary>
        /// Xuat ra dinh dang ASS - chi xuat cac dong Dialogue, khong xuat header
        /// </summary>
        private static string ToAssText(List<SubtitleLine> lines)
        {
            var sb = new StringBuilder();

            foreach (var line in lines)
            {
                var assLine = line as AssSubtitleLine;
                if (assLine != null)
                {
                    sb.AppendLine(assLine.RebuildLine());
                }
                else
                {
                    // Neu la SrtSubtitleLine, chuyen doi sang format ASS
                    var srtLine = line as SrtSubtitleLine;
                    if (srtLine != null)
                    {
                        var dialogueLine = $"Dialogue: 0,{SubtitleLine.FormatAssTime(line.StartTime)},{SubtitleLine.FormatAssTime(line.EndTime)},Default,,0000,0000,0000,,{srtLine.Text.Replace(Environment.NewLine, "\\N")}";
                        sb.AppendLine(dialogueLine);
                    }
                }
            }

            return sb.ToString().TrimEnd();
        }

        /// <summary>
        /// Xuat danh sach SubtitleLine thanh noi dung text giu nguyen format goc
        /// </summary>
        public static string ToOriginalText(List<SubtitleLine> lines, SubtitleFormat format)
        {
            return ToText(lines, format);
        }
    }
}
